### PR TITLE
Fixup preview issues with trashbin/versions

### LINKED ISSUE #32346: [10.0.9]Trying to get owner from an erased file
After upgrading from 10.0.8 to 10.0.9 I'm getting a lot of this errors:

```
{"reqId":"aVZ86bwUZOHjbSkPtaCu","level":3,"time":"2018-08-14T17:45:07+02:00","remoteAddr":"","user":"--","app":"no app in context","method":"--","url":"--","message":"Exception: {"Exception":"OCP\Files\NotFoundException","Message":"/trabajos/xxxxxxs.xlsx.v1526485490 not found while trying to get owner","Code":0,"Trace":"
#0 /var/www/owncloud/lib/private/Files/Filesystem.php(960): OC\Files\View->getOwner('/trabajos/aries...')
#1 /var/www/owncloud/lib/private/Preview.php(1315): OC\Files\Filesystem::getOwner('/trabajos/xxxxx...')
#2 /var/www/owncloud/lib/private/Preview.php(1303): OC\Preview::post_delete(Array, 'files/')
#3 /var/www/owncloud/lib/private/legacy/hook.php(105): OC\Preview::post_delete_versions(Array)
#4 /var/www/owncloud/apps/files_versions/lib/Storage.php(504): OC_Hook::emit('\OCP\Versions', 'delete', Array)
#5 /var/www/owncloud/apps/files_versions/lib/BackgroundJob/ExpireVersions.php(75): OCA\Files_Versions\Storage::expireOlderThanMaxForUser('lizzardts')
#6 /var/www/owncloud/lib/private/User/Manager.php(440): OCA\Files_Versions\BackgroundJob\ExpireVersions->OCA\Files_Versions\BackgroundJob\{closure}(Object(OC\User\User))
#7 /var/www/owncloud/lib/private/User/AccountMapper.php(233): OC\User\Manager->OC\User\{closure}(Object(OC\User\Account))
#8 /var/www/owncloud/lib/private/User/Manager.php(441): OC\User\AccountMapper->callForAllUsers(Object(Closure), '', true)
#9 /var/www/owncloud/lib/private/User/Manager.php(459): OC\User\Manager->callForAllUsers(Object(Closure), '', true)
#10 /var/www/owncloud/apps/files_versions/lib/BackgroundJob/ExpireVersions.php(76): OC\User\Manager->callForSeenUsers(Object(Closure))
#11 /var/www/owncloud/lib/private/BackgroundJob/Job.php(57): OCA\Files_Versions\BackgroundJob\ExpireVersions->run(NULL)
#12 /var/www/owncloud/lib/private/BackgroundJob/TimedJob.php(53): OC\BackgroundJob\Job->execute(Object(OC\BackgroundJob\JobList), Object(OC\Log))
#13 /var/www/owncloud/cron.php(121): OC\BackgroundJob\TimedJob->execute(Object(OC\BackgroundJob\JobList), Object(OC\Log))
#14 {main}","File":"/var/www/owncloud/lib/private/Files/View.php","Line":1718}"}
```

It doesn't look like a big error, but the log is full of those messages

### PR DESCRIPTION
**Pull request by VicDeo**
## Description
- `uid` is always known - pass it within the `version_delete` hook info
- pass `original_path` that contains a path to the versioned file 
and `deleted_revision` - versionId for the deleted version 
- pass an optional `versionId` to deleteAllPreviews when we need to remove previous for the specific version only
- bonus 1: fix preview generation for files in deleted directories
- bonus 2: cover some parts of files_versions with tests

## Related Issue
- Fixes #32346
https://github.com/owncloud/enterprise/issues/2904

## Motivation and Context
- Version previews are not deleted properly
- Previews don't show for files in deleted directories
- Versions and trash expiration is flooding the log file with exceptions related to the previews

### Case 1
1. Create a text file "test.txt"
2. Edit the file three times in the text editor and save each time, will create three versions
3. Open the versions panel to have it generate previews
4. Run cron.php and observe the logs

#### Expected result
version previews for cleaned up versions are deleted
No errors

#### Actual result
version previews for cleaned up versions are NOT deleted

```
{"reqId":"viyF6aTvAF41QBIxIu9B","level":3,"time":"2018-10-15T19:37:42+00:00","remoteAddr":"","user":"--","app":"no app in context","method":"--","url":"--","message":"Exception: {\"Exception\":\"OCP\\\\Files\\\\NotFoundException\",\"Message\":\"\\\/test.txt.v1539632201 not found while trying to get owner\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Filesystem.php(961): OC\\\\Files\\\\View->getOwner('\\\/test.txt.v1539...')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Preview.php(1241): OC\\\\Files\\\\Filesystem::getOwner('\\\/test.txt.v1539...')\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/legacy\\\/hook.php(105): OC\\\\Preview::prepare_delete(Array)\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_versions\\\/lib\\\/Storage.php(765): OC_Hook::emit('\\\\\\\\OCP\\\\\\\\Versions', 'preDelete', Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_versions\\\/lib\\\/Command\\\/Expire.php(60): OCA\\\\Files_Versions\\\\Storage::expire('\\\/test.txt', 'admin')\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Command\\\/CommandJob.php(34): OCA\\\\Files_Versions\\\\Command\\\\Expire->handle()\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/BackgroundJob\\\/Job.php(57): OC\\\\Command\\\\CommandJob->run('O:33:\\\"OCA\\\\\\\\Files...')\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/BackgroundJob\\\/QueuedJob.php(42): OC\\\\BackgroundJob\\\\Job->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/cron.php(120): OC\\\\BackgroundJob\\\\QueuedJob->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#9 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/View.php\",\"Line\":1717}"}
{"reqId":"viyF6aTvAF41QBIxIu9B","level":3,"time":"2018-10-15T19:37:42+00:00","remoteAddr":"","user":"--","app":"no app in context","method":"--","url":"--","message":"Exception: {\"Exception\":\"OCP\\\\Files\\\\NotFoundException\",\"Message\":\"\\\/test.txt.v1539632201 not found while trying to get owner\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Filesystem.php(961): OC\\\\Files\\\\View->getOwner('\\\/test.txt.v1539...')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Preview.php(1311): OC\\\\Files\\\\Filesystem::getOwner('\\\/test.txt.v1539...')\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Preview.php(1299): OC\\\\Preview::post_delete(Array, 'files\\\/')\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/legacy\\\/hook.php(105): OC\\\\Preview::post_delete_versions(Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_versions\\\/lib\\\/Storage.php(767): OC_Hook::emit('\\\\\\\\OCP\\\\\\\\Versions', 'delete', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_versions\\\/lib\\\/Command\\\/Expire.php(60): OCA\\\\Files_Versions\\\\Storage::expire('\\\/test.txt', 'admin')\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Command\\\/CommandJob.php(34): OCA\\\\Files_Versions\\\\Command\\\\Expire->handle()\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/BackgroundJob\\\/Job.php(57): OC\\\\Command\\\\CommandJob->run('O:33:\\\"OCA\\\\\\\\Files...')\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/BackgroundJob\\\/QueuedJob.php(42): OC\\\\BackgroundJob\\\\Job->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/cron.php(120): OC\\\\BackgroundJob\\\\QueuedJob->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#10 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/View.php\",\"Line\":1717}"}
```

### Case 2
1. create a text files, edit it several times 
2. open versions to create a previews for versions
3. edit file one more time

#### Expected result
Only file previews are deleted

#### Actual result
File previews  and all versions previews are deleted

### Case 3
delete a directory with a file and browse to the file in the trashbin

#### Expected result
File in deleted directory has a prievew

#### Actual result
No file preview

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [x] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

File: lib/private/Preview.php [hunk 1]
```diff
@@ -415,6 +417,10 @@ public function deleteAllPreviews() {
 				// .ocTransferId*.part file from chunked file upload.
 				if (!empty($fileId)) {
 					$previewPath = $this->getPreviewPath($fileId);
+					if ($versionId !== null) {
+						$versionId = (int) $versionId;
```
Comment (PVince81) [inline lib/private/Preview.php]: bigint?

File: lib/private/Preview.php [hunk 2]
```diff
@@ -1300,7 +1306,19 @@ public static function post_delete_files($args) {
...
 	 */
 	public static function post_delete_versions($args) {
-		self::post_delete($args, 'files/');
+		$versionArgs = $args;
+		// Are we doing a rollback?
+		if (isset($args['revision'])) {
+			self::post_delete($args, 'files/');
+		} else {
+			// split path and timestamp
+			\preg_match_all('#(?<path>.*)\.v(?<timestamp>\d+)$#', $versionArgs['path'], $versionInfo);
```
Comment (PVince81) [inline lib/private/Preview.php]: put this in file helper as well ?

**Review by PVince81 — COMMENTED** [review]
See comments.

I wonder if this is enough as there might be other code paths related to trashbin: try having version previews for files that were deleted in trashbin, the trigger trashbin expiration. This will like trigger https://github.com/owncloud/enterprise/issues/3067 and might trigger some other hook methods.

**Comment by PVince81** [comment]
I like how you refactored the static methods to be able to write tests!

**Comment by PVince81** [comment]
codecov however seems less enthusiastic... and fails to load :rage:

File: lib/private/Preview.php (hunk 2 above)
Comment (VicDeo) [reply lib/private/Preview.php]: I don't like the idea of using a helper from app in core directly.

Comment (VicDeo) [reply lib/private/Preview.php]: maybe this part should be moved into the files_versions and I just add an extra params to the $args

Comment (PVince81) [reply lib/private/Preview.php]: ah, indeed. I didn't notice that the helper was in the app and this was core.

never mind then

File: lib/private/Preview.php (hunk 1 above)
Comment (VicDeo) [reply lib/private/Preview.php]: thanks. intval is better  as it casts to int/bigint depending on the platform

File: lib/private/Preview.php (hunk 2 above)
Comment (VicDeo) [reply lib/private/Preview.php]: moved into to the hook params

**Comment by VicDeo** [comment]
>codecov however seems less enthusiastic... and fails to load 

@PVince81 
it annoys me the most because I need to see what else can be covered there instead of using :crystal_ball:  and reading the starts for that purpose

**Comment by VicDeo** [comment]
> I wonder if this is enough as there might be other code paths related to trashbin: try having version previews for files that were deleted in trashbin, the trigger trashbin expiration. This will like trigger owncloud/enterprise#3067 and might trigger some other hook methods.

@PVince81 
I tried to delete the test file after steps 1-3 from the PR description
**Results**: 
- directory `data/D/thumbnails/2147496457/` was deleted so no preview of any kind exists anymore. 
- after browsing to trashbin previews for the file `data/D/thumbnails/2147496457/`  were recreated
previews for versions were NOT recreated after viewing the file in trashbin.

here is log (nothing fishy)
```
{"reqId":"952b0382-c5d9-4f17-904e-ebb3647cb83d","level":1,"time":"2019-02-01T16:34:35+00:00","remoteAddr":"127.0.0.1","user":"D","app":"admin_audit","method":"GET","url":"\/~deo\/oc-tmp\/remote.php\/dav\/files\/D\/bebe.txt","message":"Read \"\/D\/files\/bebe.txt\" by user <D>, displayName <Dryndyndyn>, owner: D [CLIENT_IP: 127.0.0.1] [USER_AGENT: Mozilla\/5.0 (Linux) mirall\/2.5.2 (build 268)]"}
{"reqId":"uZHeLxVvFDf2HZX1gS4U","level":0,"time":"2019-02-01T16:35:04+00:00","remoteAddr":"127.0.0.1","user":"D","app":"core","method":"GET","url":"\/~deo\/oc-tmp\/remote.php\/dav\/meta\/2147496457\/v\/1549038093?preview","message":"Generating preview for \"\/meta\/2147496457\/v\/1549038093\" with \"OC\\Preview\\TXT\""}
{"reqId":"S6IUMO4v2hUmT44uHYvy","level":0,"time":"2019-02-01T16:35:04+00:00","remoteAddr":"127.0.0.1","user":"D","app":"core","method":"GET","url":"\/~deo\/oc-tmp\/remote.php\/dav\/meta\/2147496457\/v\/1549038077?preview","message":"Generating preview for \"\/meta\/2147496457\/v\/1549038077\" with \"OC\\Preview\\TXT\""}
{"reqId":"qxEPyFCMFNDb25iyDUsk","level":0,"time":"2019-02-01T16:35:05+00:00","remoteAddr":"127.0.0.1","user":"D","app":"core","method":"GET","url":"\/~deo\/oc-tmp\/remote.php\/dav\/meta\/2147496457\/v\/1549038084?preview","message":"Generating preview for \"\/meta\/2147496457\/v\/1549038084\" with \"OC\\Preview\\TXT\""}
{"reqId":"VxXof4wusgbt9efZIhKO","level":1,"time":"2019-02-01T16:35:18+00:00","remoteAddr":"127.0.0.1","user":"D","app":"admin_audit","method":"DELETE","url":"\/~deo\/oc-tmp\/remote.php\/dav\/files\/D\/bebe.txt","message":"Delete \"\/D\/files\/bebe.txt\" by user <D>, displayName <Dryndyndyn>, owner: D [CLIENT_IP: 127.0.0.1] [USER_AGENT: Mozilla\/5.0 (X11; Linux x86_64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/72.0.3626.81 Safari\/537.36]"}
{"reqId":"6RtK4Es9C3bSFanQm7sF","level":0,"time":"2019-02-01T16:40:48+00:00","remoteAddr":"127.0.0.1","user":"D","app":"core","method":"GET","url":"\/~deo\/oc-tmp\/index.php\/apps\/files_trashbin\/ajax\/preview.php?file=%2Fbebe.txt.d1549038918&c=1549038918000","message":"Generating preview for \"\/D\/files_trashbin\/files\/bebe.txt.d1549038918\" with \"OC\\Preview\\TXT\""}
```
So there is nothing to fix for the trashbin IMO unless such agressive deletion of preview is unexpected

**Comment by VicDeo** [comment]
The fix is incomplete, preDelete hook should also has the same changes.
The same applies to trashbin as for trashbin '/files' is added after userid producing weird path like 
`/D/files/files_trashbin/files/bebe.txt.d1549039401` as a result

**Comment by PVince81** [comment]
> So there is nothing to fix for the trashbin IMO unless such agressive deletion of preview is unexpected

I seem to remember that the code was deleting more than just the expected previews, which is a bit brutal. Can you recheck the exact path that is getting deleted from the code ?

**Comment by VicDeo** [comment]
@PVince81 
> the code was deleting more than just the expected previews, which is a bit brutal.

it's due `\OC\Preview::post_write` hands off the job to `\OC\Preview::post_delete` which removes all preview by default.

**Comment by PVince81** [comment]
@VicDeo right, now I remember. can we make it just remove what is expected ?

**Comment by VicDeo** [comment]
hmmm... I thought it would be easier
 To test:
- [ ] all nested files/version previews are deleted when directory is  removed
- [ ] post write hook deletes file preview only (version previews are kept)
- [ ] file/directory previews are deleted when trashbin is cleaned

**Comment by VicDeo** [comment]
@PVince81  description updated to describe 3 bugs that are already fixed in this PR. 
Only trashbin expiration is left to fix.

**Comment by VicDeo** [comment]
I'll list my expectations here. This is how the things work within this PR:

1. file preview is created when the file is opened in a file list as `/user/thumbnails/fileId/x-y.png`
2. version preview is created when the version is listed in the versions tab as `/user/thumbnails/fileId/versionId/x-y.png`
3. All file previews are deleted when file is updated: all `/user/thumbnails/fileId/*.png` are deleted
4. All version previews are kept when file is updated `/user/thumbnails/fileId/versionId/` still exists
5. File preview is deleted when the file is moved to trashbin: entire  `/user/thumbnails/fileId/` directory is deleted
6. Versions previews are deleted when the file is moved to trashbin: entire  `/user/thumbnails/fileId/` directory is deleted
7. After browsing to the file in the trashbin preview is recreated in `/user/thumbnails/fileId/`
8. After browsing to the file in the version previews are not recreated in `/user/thumbnails/fileId/`
9. After a file with a preview is deleted from trashbin - preview is deleted: entire  `/user/thumbnails/fileId/` directory is deleted
10. After a file with a preview is restored from trashbin - previews are kept

**Comment by VicDeo** [comment]
@PVince81 please rereview

File: apps/files_versions/lib/Storage.php [hunk 3]
```diff
@@ -248,9 +236,16 @@ public static function delete($path) {
 			$versions = self::getVersions($uid, $filename);
 			if (!empty($versions)) {
 				foreach ($versions as $v) {
-					\OC_Hook::emit('\OCP\Versions', 'preDelete', ['path' => $path . $v['version'], 'trigger' => self::DELETE_TRIGGER_MASTER_REMOVED]);
+					$hookData = 						[
```
Comment (PVince81) [inline apps/files_versions/lib/Storage.php]: faraway `[`

File: apps/files_versions/lib/Storage.php [hunk 4]
```diff
@@ -443,11 +438,7 @@ public static function getVersions($uid, $filename, $userFullPath = '') {
 						$key = $timestamp . '#' . $filename;
 						$versions[$key]['version'] = $timestamp;
 						$versions[$key]['humanReadableTimestamp'] = self::getHumanReadableTimestamp($timestamp);
-						if (empty($userFullPath)) {
-							$versions[$key]['preview'] = '';
-						} else {
-							$versions[$key]['preview'] = \OCP\Util::linkToRoute('core_ajax_versions_preview', ['file' => $userFullPath, 'version' => $timestamp]);
-						}
+						$versions[$key]['preview'] = '';
```
Comment (PVince81) [inline apps/files_versions/lib/Storage.php]: gets populated later on ?

File: apps/files_versions/tests/FileHelperTest.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+		$this->assertEquals(
+			[
+				'/docs/file1.txt',
+				'/file1.txt',
+			],
+			\array_keys($versionsByFile)
+		);
+	}
+
+	protected function descendByPath($path, $tree) {
```
Comment (PVince81) [inline apps/files_versions/tests/FileHelperTest.php]: phpdoc for this complex logic ?

File: lib/private/Preview.php [hunk 6]
```diff
@@ -152,9 +151,9 @@ public function __construct(
 	}
 
 	/**
-	 * returns the path of the file you want a thumbnail from
+	 * returns the node you want a thumbnail from
 	 *
-	 * @return File
+	 * @return Node | null
```
Comment (PVince81) [inline lib/private/Preview.php]: Node, so can it be a folder also ?

**Review by PVince81 — COMMENTED** [review]
some minor questions

File: apps/files_versions/lib/Storage.php (hunk 4 above)
Comment (VicDeo) [reply apps/files_versions/lib/Storage.php]: Looks like unused at all. We have no `core_ajax_versions_preview` route and I haven't heard any complains :)

Comment (PVince81) [reply apps/files_versions/lib/Storage.php]: right. that route got obliterated after moving to Webdav for non-public previews

File: lib/private/Preview.php (hunk 6 above)
Comment (VicDeo) [reply lib/private/Preview.php]: may be yes. At least  constructor accepts `Node`, not `File` https://github.com/owncloud/core/pull/34283/files/c06b2f85b1f9b733f703ccee12ca5f67ef0e674d#diff-e9181f7e8cee7044eda5833fdb98915cL115

Comment (VicDeo) [reply lib/private/Preview.php]: ..and logic tells me that folder preview could be a folder icon.

File: apps/files_versions/tests/FileHelperTest.php (hunk 5 above)
Comment (VicDeo) [reply apps/files_versions/tests/FileHelperTest.php]: added

File: apps/files_versions/lib/Storage.php (hunk 3 above)
Comment (VicDeo) [reply apps/files_versions/lib/Storage.php]: Fixed

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by PVince81** [comment]
@VicDeo please backport

**Comment by VicDeo** [comment]
stable10: https://github.com/owncloud/core/pull/34533

**Pull request merged**